components/net/websocket_loader.rs

```diff
@@ -3,63 +3,107 @@
-    };
-
-    let mut request = try!(Client::connect(net_url));
-    request.headers.set(Origin(origin));
-    request.headers.set(host);
-    if !protocols.is_empty() {
-        request.headers.set(WebSocketProtocol(protocols.clone()));
-    };
-
-    http_loader::set_request_cookies(&resource_url, &mut request.headers, &cookie_jar);
-
-    let response = try!(request.send());
-    try!(response.validate());
-
-    {
-       let protocol_in_use = unwrap_websocket_protocol(response.protocol());
-        if let Some(protocol_name) = protocol_in_use {
-                if !protocols.is_empty() && !protocols.iter().any(|p| (&**p).eq_ignore_ascii_case(protocol_name)) {
-                    return Err(WebSocketError::ProtocolError("Protocol in Use not in client-supplied protocol list"));
-            };
+use url::Url;
+use ws::{CloseCode, Factory, Handler, Handshake, Message, Request, Response, Sender, WebSocket};
+use ws::{Error as WebSocketError, ErrorKind as WebSocketErrorKind, Result as WebSocketResult};
+
+/// A client for connecting to a websocket server
+#[derive(Clone, Copy)]
+struct Client<'a> {
+    origin: &'a str,
+    host: &'a Host,
+    protocols: &'a [String],
+    cookie_jar: &'a Arc<RwLock<CookieStorage>>,
+    resource_url: &'a ServoUrl,
+    event_sender: &'a IpcSender<WebSocketNetworkEvent>,
+}
+
+impl<'a> Factory for Client<'a> {
+    type Handler = Self;
+
+    fn connection_made(&mut self, _: Sender) -> Self::Handler {
+        *self
+    }
+
+    fn connection_lost(&mut self, _: Self::Handler) {
+        let _ = self.event_sender.send(WebSocketNetworkEvent::Fail);
+    }
+}
+
+impl<'a> Handler for Client<'a> {
+    fn build_request(&mut self, url: &Url) -> WebSocketResult<Request> {
+        debug!("Handler is building request from {}.", url);
+        try!(parse_url(url).map_err(|e| WebSocketError::new(WebSocketErrorKind::Protocol, e.description().to_owned())));
+        let mut req = try!(Request::from_url(url));
+        req.headers_mut().push(("Origin".to_string(), self.origin.as_bytes().to_owned()));
+        req.headers_mut().push(("Host".to_string(), format!("{}", self.host).as_bytes().to_owned()));
+        for protocol in self.protocols {
+            req.add_protocol(protocol);
         };
+        let mut headers = WebSocketHeaders::new(req.headers().clone()).as_hyper_headers();
+        set_request_cookies(self.resource_url, &mut headers, self.cookie_jar);
+        if let Some(cookie) = headers.get_raw("Cookie") {
+            req.headers_mut().push(("Cookie".to_string(), cookie[0].to_owned()));
+        }
+        Ok(req)
     }
 
-    let headers = response.headers.clone();
-    let (sender, receiver) = response.begin().split();
-    Ok((headers, sender, receiver))
+    fn on_open(&mut self, shake: Handshake) -> WebSocketResult<()> {
+        let headers = WebSocketHeaders::new(shake.response.headers().clone());
+        set_cookies_from_headers(self.resource_url, &headers.as_hyper_headers(), self.cookie_jar);
```

Comment:
This changes the current behaviour and goes against the spec - cookies are set from the queued task, rather than immediately on connection: https://dxr.mozilla.org/servo/rev/e8b364793416465f6b4877414623dac2c52f2e6a/components/script/dom/websocket.rs#500-505